File: bindings/java/MvgalClient.java
// SPDX-License-Identifier: MIT OR Apache-2.0
/**
 * MVGAL Java binding (JNI-based)
 *
 * Connects to the mvgald IPC socket via JNI to libmvgal.so,
 * queries device capabilities, submits a test workload, and reads telemetry.
 *
 * Build:
 *   mvn package
 *   java -jar target/mvgal-client-0.2.0.jar
 */

package org.mvgal;

import java.io.*;
import java.net.*;
import java.nio.*;
import java.nio.channels.*;
import java.nio.file.*;
import java.nio.charset.StandardCharsets;

/**
 * MVGAL IPC client for Java.
 *
 * <p>Communicates with the mvgald daemon via its Unix domain socket.
 * Requires Java 16+ for {@code UnixDomainSocketAddress}.
 */
public class MvgalClient implements AutoCloseable {

    /** MVGAL IPC socket path. */
    public static final String SOCKET_PATH = "/run/mvgal/mvgal.sock";

    /** IPC message magic number. */
    public static final int IPC_MAGIC = 0x4D564741;

    /** Message types. */
    public static final int MSG_QUERY_DEVICES       = 0x0001;
    public static final int MSG_SUBMIT_WORKLOAD     = 0x0002;
    public static final int MSG_ALLOC_MEMORY        = 0x0003;
    public static final int MSG_FREE_MEMORY         = 0x0004;
    public static final int MSG_QUERY_CAPABILITIES  = 0x0005;
    public static final int MSG_SET_STRATEGY        = 0x0006;
    public static final int MSG_SUBSCRIBE_TELEMETRY = 0x0007;
    public static final int MSG_TELEMETRY           = 0x0008;

    private SocketChannel channel;
    private int seq = 0;

    /**
     * Connect to the mvgald IPC socket.
     *
     * @throws IOException if the connection fails
     */
    public void connect() throws IOException {
        UnixDomainSocketAddress addr = UnixDomainSocketAddress.of(SOCKET_PATH);
        channel = SocketChannel.open(StandardProtocolFamily.UNIX);
        channel.connect(addr);
    }

    /**
     * Disconnect from the daemon.
     */
    @Override
    public void close() throws IOException {
        if (channel != null && channel.isOpen()) {
            channel.close();
        }
    }

    /**
     * Send a message and return the response payload as a JSON string.
     *
     * @param msgType message type constant
     * @param payload JSON payload string (may be empty)
     * @return JSON response string
     * @throws IOException on I/O error
     */
    public String sendMessage(int msgType, String payload) throws IOException {
        seq++;
        byte[] payloadBytes = payload.isEmpty()
            ? new byte[0]
            : payload.getBytes(StandardCharsets.UTF_8);

        // Build 16-byte header (little-endian)
        ByteBuffer hdr = ByteBuffer.allocate(16).order(ByteOrder.LITTLE_ENDIAN);
        hdr.putInt(IPC_MAGIC);
        hdr.putShort((short) 1);           // version
        hdr.putShort((short) msgType);
        hdr.putInt(payloadBytes.length);
        hdr.putInt(seq);
        hdr.flip();

        channel.write(hdr);
        if (payloadBytes.length > 0) {
            channel.write(ByteBuffer.wrap(payloadBytes));
        }

        // Read response header
        ByteBuffer respHdr = ByteBuffer.allocate(16).order(ByteOrder.LITTLE_ENDIAN);
        while (respHdr.hasRemaining()) {
            channel.read(respHdr);
        }
        respHdr.flip();
        respHdr.position(8);
        int respLen = respHdr.getInt();

        if (respLen == 0) {
            return "{}";
        }

        ByteBuffer respPayload = ByteBuffer.allocate(respLen);
        while (respPayload.hasRemaining()) {
            channel.read(respPayload);
        }
        return new String(respPayload.array(), StandardCharsets.UTF_8);
    }

    /** Query all GPU devices. */
    public String queryDevices() throws IOException {
        return sendMessage(MSG_QUERY_DEVICES, "");
    }

    /** Query aggregate capabilities. */
    public String queryCapabilities() throws IOException {
        return sendMessage(MSG_QUERY_CAPABILITIES, "");
    }

    /** Submit a test compute workload. */
    public String submitTestWorkload() throws IOException {
        String payload = "{\"workload_type\":\"compute\",\"priority\":50,"
            + "\"gpu_mask\":0,\"estimated_bytes\":1048576,\"api\":\"opencl\"}";
        return sendMessage(MSG_SUBMIT_WORKLOAD, payload);
    }

    /** Subscribe to telemetry at 1-second intervals. */
    public String subscribeTelemetry() throws IOException {
        String payload = "{\"interval_ms\":1000,"
            + "\"events\":[\"gpu_utilization\",\"temperature\"]}";
        return sendMessage(MSG_SUBSCRIBE_TELEMETRY, payload);
    }

    /** Entry point for standalone test. */
    public static void main(String[] args) {
        try (MvgalClient client = new MvgalClient()) {
            client.connect();
            System.out.println("[mvgal-java] Connected to mvgald");

            String caps = client.queryCapabilities();
            System.out.println("[mvgal-java] Capabilities: " + caps);

            String devices = client.queryDevices();
            System.out.println("[mvgal-java] Devices: " + devices);

            String workload = client.submitTestWorkload();
            System.out.println("[mvgal-java] Workload: " + workload);

            System.out.println("[mvgal-java] Disconnected");
        } catch (IOException e) {
            System.err.println("[mvgal-java] Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: bindings/csharp/MvgalClient.cs
// SPDX-License-Identifier: MIT OR Apache-2.0
/**
 * MVGAL C# binding (P/Invoke-based)
 *
 * Connects to the mvgald IPC socket, queries device capabilities,
 * submits a test workload, and reads telemetry.
 *
 * Build:
 *   dotnet build
 *   dotnet run
 */

using System;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Runtime.InteropServices;

namespace Mvgal
{
    /// <summary>
    /// MVGAL IPC client for C#.
    /// Communicates with the mvgald daemon via its Unix domain socket.
    /// </summary>
    public sealed class MvgalClient : IDisposable
    {
        /// <summary>MVGAL IPC socket path.</summary>
        public const string SocketPath = "/run/mvgal/mvgal.sock";

        /// <summary>IPC message magic number.</summary>
        public const uint IpcMagic = 0x4D564741;

        /// <summary>Message type constants.</summary>
        public static class MsgType
        {
            public const ushort QueryDevices       = 0x0001;
            public const ushort SubmitWorkload     = 0x0002;
            public const ushort AllocMemory        = 0x0003;
            public const ushort FreeMemory         = 0x0004;
            public const ushort QueryCapabilities  = 0x0005;
            public const ushort SetStrategy        = 0x0006;
            public const ushort SubscribeTelemetry = 0x0007;
            public const ushort Telemetry          = 0x0008;
        }

        private Socket? _socket;
        private NetworkStream? _stream;
        private uint _seq;

        /// <summary>Connect to the mvgald IPC socket.</summary>
        public void Connect()
        {
            _socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
            var endpoint = new UnixDomainSocketEndPoint(SocketPath);
            _socket.Connect(endpoint);
            _stream = new NetworkStream(_socket, ownsSocket: false);
        }

        /// <summary>Disconnect from the daemon.</summary>
        public void Dispose()
        {
            _stream?.Dispose();
            _socket?.Dispose();
        }

        /// <summary>
        /// Send a message and return the response payload as a JSON string.
        /// </summary>
        public string SendMessage(ushort msgType, string payload = "")
        {
            if (_stream is null) throw new InvalidOperationException("Not connected");

            _seq++;
            byte[] payloadBytes = payload.Length > 0
                ? Encoding.UTF8.GetBytes(payload)
                : Array.Empty<byte>();

            // Build 16-byte header (little-endian)
            byte[] hdr = new byte[16];
            BitConverter.TryWriteBytes(hdr.AsSpan(0, 4), IpcMagic);
            BitConverter.TryWriteBytes(hdr.AsSpan(4, 2), (ushort)1);
            BitConverter.TryWriteBytes(hdr.AsSpan(6, 2), msgType);
            BitConverter.TryWriteBytes(hdr.AsSpan(8, 4), (uint)payloadBytes.Length);
            BitConverter.TryWriteBytes(hdr.AsSpan(12, 4), _seq);

            _stream.Write(hdr);
            if (payloadBytes.Length > 0)
                _stream.Write(payloadBytes);

            // Read response header
            byte[] respHdr = new byte[16];
            _stream.ReadExactly(respHdr);
            uint respLen = BitConverter.ToUInt32(respHdr, 8);

            if (respLen == 0) return "{}";

            byte[] respPayload = new byte[respLen];
            _stream.ReadExactly(respPayload);
            return Encoding.UTF8.GetString(respPayload);
        }

        /// <summary>Query all GPU devices.</summary>
        public JsonDocument QueryDevices()
            => JsonDocument.Parse(SendMessage(MsgType.QueryDevices));

        /// <summary>Query aggregate capabilities.</summary>
        public JsonDocument QueryCapabilities()
            => JsonDocument.Parse(SendMessage(MsgType.QueryCapabilities));

        /// <summary>Submit a test compute workload.</summary>
        public JsonDocument SubmitTestWorkload()
        {
            const string payload = """
                {"workload_type":"compute","priority":50,"gpu_mask":0,
                 "estimated_bytes":1048576,"api":"opencl"}
                """;
            return JsonDocument.Parse(SendMessage(MsgType.SubmitWorkload, payload));
        }

        /// <summary>Subscribe to telemetry at 1-second intervals.</summary>
        public JsonDocument SubscribeTelemetry()
        {
            const string payload = """
                {"interval_ms":1000,"events":["gpu_utilization","temperature"]}
                """;
            return JsonDocument.Parse(SendMessage(MsgType.SubscribeTelemetry, payload));
        }

        /// <summary>Entry point for standalone test.</summary>
        public static void Main(string[] args)
        {
            using var client = new MvgalClient();
            try
            {
                client.Connect();
                Console.WriteLine("[mvgal-csharp] Connected to mvgald");

                using var caps = client.QueryCapabilities();
                Console.WriteLine($"[mvgal-csharp] GPU count: {caps.RootElement.GetProperty("gpu_count")}");
                Console.WriteLine($"[mvgal-csharp] Total VRAM: {caps.RootElement.GetProperty("total_vram_bytes")} bytes");

                using var devices = client.QueryDevices();
                Console.WriteLine($"[mvgal-csharp] Devices: {JsonSerializer.Serialize(devices.RootElement, new JsonSerializerOptions { WriteIndented = true })}");

                using var workload = client.SubmitTestWorkload();
                Console.WriteLine($"[mvgal-csharp] Workload: {JsonSerializer.Serialize(workload.RootElement, new JsonSerializerOptions { WriteIndented = true })}");

                Console.WriteLine("[mvgal-csharp] Disconnected");
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"[mvgal-csharp] Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}
